src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeploymentPosterImpl.java

```diff
@@ -10,43 +10,105 @@
 import com.atlassian.bitbucket.jenkins.internal.credentials.JenkinsToBitbucketCredentials;
 import com.atlassian.bitbucket.jenkins.internal.model.deployment.BitbucketCDCapabilities;
 import com.atlassian.bitbucket.jenkins.internal.model.deployment.BitbucketDeployment;
+import com.atlassian.bitbucket.jenkins.internal.model.deployment.BitbucketDeploymentEnvironment;
+import com.atlassian.bitbucket.jenkins.internal.provider.JenkinsProvider;
+import com.atlassian.bitbucket.jenkins.internal.scm.BitbucketSCMRepository;
+import com.atlassian.bitbucket.jenkins.internal.scm.BitbucketSCMRepositoryHelper;
 import com.cloudbees.plugins.credentials.Credentials;
+import hudson.Extension;
+import hudson.FilePath;
+import hudson.model.FreeStyleBuild;
 import hudson.model.Run;
 import hudson.model.TaskListener;
+import hudson.model.listeners.SCMListener;
+import hudson.plugins.git.GitSCM;
+import hudson.scm.SCM;
+import hudson.scm.SCMRevisionState;
+import hudson.tasks.Publisher;
 
+import javax.annotation.CheckForNull;
 import javax.inject.Inject;
 import javax.inject.Singleton;
+import java.io.File;
+import java.io.IOException;
 import java.util.Optional;
 import java.util.logging.Level;
 import java.util.logging.Logger;
 
 import static java.lang.String.format;
+import static org.apache.commons.lang3.StringUtils.isBlank;
 
+@Extension
 @Singleton
-public class DeploymentPosterImpl implements DeploymentPoster {
+public class DeploymentPosterImpl extends SCMListener implements DeploymentPoster {
 
     protected static final Logger LOGGER = Logger.getLogger(DeploymentPosterImpl.class.getName());
 
-    private final BitbucketClientFactoryProvider bitbucketClientFactoryProvider;
-    private final JenkinsToBitbucketCredentials jenkinsToBitbucketCredentials;
-    private final BitbucketPluginConfiguration pluginConfiguration;
+    private BitbucketClientFactoryProvider bitbucketClientFactoryProvider;
+    private BitbucketDeploymentFactory bitbucketDeploymentFactory;
+    private JenkinsProvider jenkinsProvider;
+    private JenkinsToBitbucketCredentials jenkinsToBitbucketCredentials;
+    private BitbucketPluginConfiguration pluginConfiguration;
+    private BitbucketSCMRepositoryHelper scmRunHelper;
+
+    public DeploymentPosterImpl() {
+
```

Comment:
I am not a fan of this double constructor business we've got going here. I realise that because you're an `@Extension` there's a compiler check to enforce you have this no args constructor. 
Maybe setter injection would be better? That way it is clearer to us humans what is going on?
@mhenschke-atl  what do you think?

Reply:
Again for posterity- neither option is clean, but we should never instantiate an extension. We can mock them in unit tests, and pull the instance out of Jenkins in integration tests. This is a common enough pattern that I think either approach is fine.